package/MDAnalysis/analysis/hbond_analysis.py

```diff
@@ -0,0 +1,204 @@
+
+Hydrogen bond data are returned in a :class:`numpy.ndarray` on a "one line, one observation" basis
+and can be accessed via :attr:`HydrogenBondAnalysis.hbonds`::
+
+    results = [
+        [
+            <frame>,
+            <donor index (0-based)>,
+            <hydrogen index (0-based)>,
+            <acceptor index (0-based)>,
+            <distance>,
+            <angle>
+        ],
+        ...
+    ]
+
+Classes
+-------
+
+.. autoclass:: HydrogenBondAnalysis
+   :members:
+
+   .. attribute:: hbonds
+
+"""
+from __future__ import absolute_import
+
+import numpy as  np
+
+from . import base
+from MDAnalysis.lib.distances import capped_distance, calc_angles
+
+class HydrogenBondAnalysis(base.AnalysisBase):
+    """
+    Perform an analysis of hydrogen bonds in a Universe.
+    """
+
+    def __init__(self, universe, donors_sel, hydrogens_sel, acceptors_sel, d_h_cutoff, d_a_cutoff, d_h_a_angle_cutoff):
+        """Set up atom selections and geometric criteria for finding hydrogen bonds in a Universe.
+
+        Parameters
+        ----------
+        universe : Universe
+            MDAnalysis Universe object
+        donors_sel : str
+            Selection string for the hydrogen bond donor atoms
+        hydrogens_sel :  str
+            Selection string for the hydrogen bond hydrogen atoms
+        acceptors_sel : str
+            Selection string for the hydrogen bond acceptor atoms
+        d_h_cutoff : float (optional)
+            Distance cutoff used for finding donor-hydrogen pairs [1.2]
+        d_a_cutoff : float (optional)
+            Distance cutoff for hydrogen bonds. This cutoff refers to the D-A distance. [3.0]
+        d_h_a_angle_cutoff: float (optional)
+            D-H-A angle cutoff for hydrogen bonds, in degrees. [150]
+
+        """
+
+        self.u = universe
+        self._trajectory = self.u.trajectory
+        self.donors_sel = donors_sel
+        self.hydrogens_sel = hydrogens_sel
+        self.acceptors_sel = acceptors_sel
+        self.d_h_cutoff = d_h_cutoff
+        self.d_a_cutoff = d_a_cutoff
+        self.d_h_a_angle = d_h_a_angle_cutoff
+
+    def _prepare(self):
+        self.hbonds = [[], [], [], [], [], []]
+        self.frames = np.arange(self.start, self.stop, self.step)
+        self.timesteps = self.frames * self.u.trajectory.dt + self.u.trajectory[self.start].time
+
+    def _single_frame(self):
+
+        box = self._ts.dimensions
+
+        # Find D-H pairs
+        all_donors = self.u.select_atoms(self.donors_sel)
+        all_hydrogens = self.u.select_atoms(self.hydrogens_sel)
+        donors_indices, hydrogen_indices = capped_distance(
+            all_donors.positions,
+            all_hydrogens.positions,
+            max_cutoff=self.d_h_cutoff,
+            box=box,
+            return_distances=False
+        ).T
+
+        # Select donors, hydrogens and acceptors
+        donors = all_donors[donors_indices]
+        hydrogens = all_hydrogens[hydrogen_indices]
```

Comment:
What would be the best way to find hydrogens via the topology? It's not something I'm very familiar with.

Reply:
I just checked this in the test WaterPSF system and, if selecting all donors in the system, it is 10 times faster to select both groups and then use the `capped_distance()` method. However, if the donors and hydrogens selections are more complicated (such as involving `and around 5 resid 7`) then the distance based method is around 40% faster than using the topology. So as the complexity of the selection increases it may become faster to use the topology to find hydrogens, but generally it is faster using `capped_distances()`.

Reply:
In principle, selecting via topology is the only _correct_ way of finding the bonded hydrogens (assuming that the bonds were not guessed). 

Do you have to find the hydrogens at every time step or can you do it for once and for all?